README.md

```diff
@@ -1,5 +1,213 @@
+The definition looks like this:
+
+```turtle
+<#pipeline> a p:Pipeline ;
+  p:variables [
+    p:variable [ a p:Variable ;
+      p:name "filename" ;
+      p:value "example.txt"
+    ]
+  ] .
+```
+
+#### Definition via Command Line Arguments
+
+Not all variables used in the pipeline need to be part of the pipeline definition.
+These abstract pipelines can use variables provided by the runtime, like the cli tool.
+In the cli tool, this can be done via the `--variable` argument.
+See the usage for more details.
+The variable from the previous example could be defined like this in the cli:
+
+`barnard59 run pipeline.json --pipeline http://example.org/#pipeline --variable filename=example.txt`    
+
+#### Using Variables via Datatype
+
+The values of the variable can be use via a literal with a `p:VariableName` datatype.
+In the example below, the argument will be replaced with the value of the variable `filename`: 
+
+```turtle
+<#readFile> a p:Step;
+  code:implementedBy [ a code:EcmaScript ;
+    code:link <node:fs#createReadStream>    
+  ] ;
+  code:arguments ("filename"^^p:VariableName) .
+```
+
+#### Using Variables in Template Literals
+
+Variables can be also used in Template Literals.
+The variable values are injected with their corresponding name, like shown in the following example: 
+
+```turtle
+<#readFile> a p:Step ;
+  code:implementedBy [ a code:EcmaScript ;
+    code:link <node:fs#createReadStream>
+  ] ;
+  code:arguments ( "${basePath}/input.txt"^^code:EcmaScriptTemplateLiteral ) .
+```
+
+#### Using Variables in ECMAScript Code
+
+Variables are also injected into the context provided to ECMAScript code.
+The context can be accessed in the standard JavaScript way using `this`.
+The variables are attached in a `Set` as `this.variables`.
+In the following example you can see how the variable `prefix` can be accessed in the JavaScript code: 
+
+```turtle
+<#map> a p:Step ;
+  code:implementedBy [ a code:EcmaScript ;
+    code:link <node:barnard59-base#map>
+  ] ;
+  code:arguments ("""row => {
+    return this.variables.get('prefix') + row
+  }"""^^code:EcmaScript) .
+```
+
+### JavaScript Arguments
+
+Part of the pipeline concept are JavaScript code arguments where describing a logic in a declarative way would be very complex, but is easy to code in one or few lines of JavaScript code.
+Also for this usage the Code Ontology is used. 
+The [RDF Code Loader package](https://github.com/zazuko/rdf-native-loader-code) provides this functionality.
+See the docs of the package for more details.
+
+#### EcmaScript
+
+Here is a small example for a step which is using the `barnard59-base#map` operation:  
+
+```turtle
+<#map> a p:Step ;
+  code:implementedBy [ a code:EcmaScript ;
+    code:link <node:barnard59-base#map>
+  ] ;
+  code:arguments ("""row => {
+    // replace all ö umlauts with oe
+    return row.split('ö').join('oe')
+  }"""^^code:EcmaScript) .
+```
+
+#### EcmaScript Template Literal
```

Comment:
This section kind of repeats `Using Variables in Template Literals`

Reply:
I know. Made the structure, then filled it. Somehow it should be there as separate topic. Maybe we fill it in the future with more stuff how the template literals can be useful and point from the variable section to this one? Maybe even reduce the stuff in the variable section?

Reply:
m'kay. let's keep it for now.